tempesta_fw/connection.c

```diff
@@ -94,6 +94,27 @@ tfw_connection_release(TfwConnection *conn)
 int
 tfw_connection_send(TfwConnection *conn, TfwMsg *msg)
 {
+	if (conn->tls) {
+		struct sk_buff *skb;
+
+		while ((skb = ss_skb_dequeue(&msg->skb_list))) {
+			int r, i;
+
+			if (skb_headlen(skb)) {
+				r = mbedtls_ssl_write(&conn->tls->ctx, skb->data, skb_headlen(skb));
```

Comment:
Basically this is the point where and why we can't live with mbedTLS and must develop our own fork. MbedTLS and Tempesta live in different paradigms. Actually mbedTLS was never developed with performance in mind (see comment https://github.com/tempesta-tech/tempesta/issues/81#issuecomment-128968333: "Probably there are communication issues (actually there are and we saw some of them) "). In particular:

1. mbedTLS doesn't care about IO while Tempesta processes packets immediately. We shouldn't adjust Tempesta code to work with mbedTLS buffers, but rather adjust mbedTLS such that it'll be aware about skb;

2. mbedTLS handles it's own states (e.g. see `ssl->state != MBEDTLS_SSL_HANDSHAKE_OVER` in `mbedtls_ssl_write()`) and does it in poor manner. Tempesta has GFSM to handle state machines efficiently.

All in all basically you need to fully redesign at least ttls/ssl_tls.c.